btx/processing/pipca.py

```diff
@@ -0,0 +1,1011 @@
+        """
+        Compute combined sample variance of two blocks
+        of data described by input parameters.
+
+        Parameters
+        ----------
+        s_n : ndarray, shape (d x 1)
+            sample variance of first block of data
+        s_m : ndarray, shape (d x 1)
+            sample variance of second block of data
+        mu_n : ndarray, shape (d x 1)
+            mean of first block of data
+        mu_m : ndarray, shape (d x 1)
+            mean of second block of data
+        n : int
+            number of observations in first block of data
+        m : int
+            number of observations in second block of data
+
+        Returns
+        -------
+        s_nm : ndarray, shape (d x 1)
+            combined sample variance of both blocks of data described by input parameters
+        """
+        s_nm = s_m
+
+        if n != 0:
+            s_nm = (
+                           ((n - 1) * s_n + (m - 1) * s_m) + (n * m * (mu_n - mu_m) ** 2) / (n + m)
+                   ) / (n + m - 1)
+
+        return s_nm
+
+    def get_model(self):
+        """
+        Method to retrieve model parameters.
+
+        Returns
+        -------
+        U_tot : ndarray, shape (d x q)
+            iPCA principal axes from model.
+        S_tot : ndarray, shape (1 x q)
+            iPCA singular values from model.
+        mu_tot : ndarray, shape (1 x d)
+            Data mean computed from all input images.
+        var_tot : ndarray, shape (1 x d)
+            Sample data variance computed from all input images.
+        """
+        if self.rank == 0:
+            U_tot = np.empty(self.num_features * self.num_components)
+            mu_tot = np.empty((self.num_features, 1))
+            var_tot = np.empty((self.num_features, 1))
+        else:
+            U_tot, mu_tot, var_tot = None, None, None
+
+        start_indices = self.split_indices[:-1]
+
+        self.comm.Gatherv(
+            self.U.flatten(),
+            [
+                U_tot,
+                self.split_counts * self.num_components,
+                start_indices * self.num_components,
+                MPI.DOUBLE,
+            ],
+            root=0,
+        )
+
+        if self.rank == 0:
+            U_tot = np.reshape(U_tot, (self.num_features, self.num_components))
+
+        self.comm.Gatherv(
+            self.mu,
+            [
+                mu_tot,
+                self.split_counts * self.num_components,
+                start_indices,
+                MPI.DOUBLE,
+            ],
+            root=0,
+        )
+        self.comm.Gatherv(
+            self.total_variance,
+            [
+                var_tot,
+                self.split_counts * self.num_components,
+                start_indices,
+                MPI.DOUBLE,
+            ],
+            root=0,
+        )
+
+        S_tot = self.S
+
+        return U_tot, S_tot, mu_tot, var_tot
+
+    def get_loss_stats(self):
+
```

Comment:
Missing docstring

Reply:
Agreed. Also the function name suggests this returns the compression loss while it only returns outliers (identified using the compression loss? please clarify)

Reply:
Fixed!